source/credentials_provider_sso.c

```diff
@@ -0,0 +1,843 @@
+    if (sso_query_context->token) {
+        aws_string_destroy_secure(sso_query_context->token);
+        sso_query_context->token = NULL;
+    }
+    sso_query_context->status_code = 0;
+    sso_query_context->error_code = 0;
+}
+
+static void s_sso_query_context_destroy(struct aws_sso_query_context *sso_query_context) {
+    if (sso_query_context == NULL) {
+        return;
+    }
+
+    s_sso_query_context_reset_request_specific_data(sso_query_context);
+    aws_byte_buf_clean_up(&sso_query_context->payload);
+    aws_byte_buf_clean_up(&sso_query_context->path_and_query);
+    aws_credentials_provider_release(sso_query_context->provider);
+    aws_retry_token_release(sso_query_context->retry_token);
+    aws_mem_release(sso_query_context->allocator, sso_query_context);
+}
+
+static struct aws_sso_query_context *s_sso_query_context_new(
+    struct aws_credentials_provider *provider,
+    aws_on_get_credentials_callback_fn callback,
+    void *user_data) {
+    struct aws_credentials_provider_sso_impl *impl = provider->impl;
+
+    struct aws_sso_query_context *sso_query_context =
+        aws_mem_calloc(provider->allocator, 1, sizeof(struct aws_sso_query_context));
+    sso_query_context->allocator = provider->allocator;
+    sso_query_context->provider = aws_credentials_provider_acquire(provider);
+    sso_query_context->original_user_data = user_data;
+    sso_query_context->original_callback = callback;
+
+    /* construct path and query */
+    struct aws_byte_cursor account_id_cursor = aws_byte_cursor_from_string(impl->sso_account_id);
+    struct aws_byte_cursor role_name_cursor = aws_byte_cursor_from_string(impl->sso_role_name);
+    struct aws_byte_cursor path_cursor = aws_byte_cursor_from_c_str("/federation/credentials?account_id=");
+    struct aws_byte_cursor role_name_param_cursor = aws_byte_cursor_from_c_str("&role_name=");
+
+    if (aws_byte_buf_init_copy_from_cursor(&sso_query_context->path_and_query, provider->allocator, path_cursor) ||
+        aws_byte_buf_append_encoding_uri_param(&sso_query_context->path_and_query, &account_id_cursor) ||
+        aws_byte_buf_append_dynamic(&sso_query_context->path_and_query, &role_name_param_cursor) ||
+        aws_byte_buf_append_encoding_uri_param(&sso_query_context->path_and_query, &role_name_cursor)) {
+        goto on_error;
+    }
+
+    if (aws_byte_buf_init(&sso_query_context->payload, provider->allocator, SSO_RESPONSE_SIZE_INITIAL)) {
+        goto on_error;
+    }
+
+    return sso_query_context;
+
+on_error:
+    s_sso_query_context_destroy(sso_query_context);
+
+    return NULL;
+}
+
+/*
+ * No matter the result, this always gets called assuming that sso_query_context is successfully allocated
+ */
+static void s_finalize_get_credentials_query(struct aws_sso_query_context *sso_query_context) {
+    struct aws_credentials *credentials = NULL;
+    if (sso_query_context->status_code == AWS_HTTP_STATUS_CODE_200_OK &&
+        sso_query_context->error_code == AWS_OP_SUCCESS) {
```

Comment:
trivial:
```suggestion
        sso_query_context->error_code == AWS_ERROR_SUCCESS) {
```

Reply:
Is it possible to be `sso_query_context->error_code == AWS_OP_SUCCESS`, but `sso_query_context->status_code` is not 200? If not, can we just check `error_code`?